pulsar-metadata/src/main/java/org/apache/pulsar/metadata/coordination/impl/ResourceLockImpl.java

```diff
@@ -188,6 +188,17 @@ private CompletableFuture<Void> acquireWithNoRevalidation(T newValue) {
         CompletableFuture<Void> result = new CompletableFuture<>();
         store.put(path, payload, Optional.of(version), EnumSet.of(CreateOption.Ephemeral))
                 .thenAccept(stat -> {
+                    if (!stat.isEphemeral()) {
```

Comment:
I hit this error from `testLookupConnectionNotCloseIfFailedToAcquireOwnershipOfBundle`, where the lock suddenly became persistent, after invalidating the cache and updating it with null.

This is a bit surprising to me that the ephemeral lock can suddenly become persistent(afaik, this is from the zk server behavior, not from zk client). I couldn't find enough zk doc/references to explain this behavior.

```java
        cache.invalidateLocalOwnerCache();
        final var lock = pulsar.getCoordinationService().getLockManager(NamespaceEphemeralData.class)
                .acquireLock(ServiceUnitUtils.path(bundle), new NamespaceEphemeralData()).join();
        lock.updateValue(null);
```

Reply:
Rings a bell, found #13066. There's probably more changes required to implement it properly in MockZooKeeper.

Reply:
There's https://github.com/apache/pulsar/blob/master/pulsar-broker/src/test/java/org/apache/pulsar/broker/MultiBrokerTestZKBaseTest.java as a way how it's currently possible to use a real Zookeeper with tests. However, adding direct support to PulsarTestContext and making it easy to override some protected method in MockedPulsarServiceBaseTest to choose it would be more flexible.

Reply:
@heesung-sn I pushed some changes to this PR for MockZooKeeperSession/MockZooKeeper to support ephemeral owner. The value for persistent nodes is now 0. Support was missing completely for multi-ops. In the mock zookeeper solution, the session id gets passed in a thread local since there's not another way.

Reply:
👆@heesung-sn do you have a chance to check what @Joforde suggested?

Reply:
I tried 2. and still see some test failures.

Reply:
Planning to get back to this work on Thursday this week.

Reply:
@heesung-sn I added `.withTestZookeeper` support to `PulsarTestContext` and an easy way to use this in tests, that's in commit 17ec37a. I modified BrokerServiceLookupTest to use it so that it runs for both. That's useful for comparing the results and catching problems when using real ZooKeeper. It seems that there's a bigger mess to fix.